### DISCUSSION TITLE
Shortcuts to add vue plugins

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#17143] [RFCs]
It is common that both users and modules to install a vue plugin. Supporting shortcuts allows a managed way to use vue plugins and also makes it easier for bridge support for module authors (vue2 uses global plugins using `Vue.use` vs vue3 using `app.use`)

Modules API:

```js
import { defineNuxtModule, addVuePlugin } from '@nuxt/kit'

export default defineNuxtModule({
  setup() {
     addVuePlugin({
       import: 'vue-gtag',
       options: {}
     })
  }
})
```

Nuxt Config API:

```js
import { defineNuxtConfig } from 'nuxt3' // or @nuxt/bridge

export default defineNuxtConfig({
  vue: {
    plugins: [
      'vue-gtag'
    ]
  }
})
```

Runtime API (nuxtApp within nuxt plugins):

```js
import { defineNuxtPlugin } from '#app'
import VueGtag from 'vue-gtag'

export default defineNuxtPlugin(() => {
  useVuePlugin(VueGVueGtagtag)
})
```

**Comment by jpsc** [comment]
This makes installing plugins more aligned with other Nuxt API's and abstracts vue 2 / vue 3 transition. I like it.

**Comment by bendwarn** [comment]
Prefer the config way for making it clear that nuxt plugins inject into nuxtApp while vue plugins inject into vue. But I am not sure whether enforcing nuxt.config in module is correct. Besides, it is weird too that nuxt modules inject into vue.
On the other hand, is it just me thinking plugins exist for compatibility when composables are so much powerful?

Btw, maybe there need to be options for server/client in config.

**Reply by pi0** [reply]
Module authors can use `addVuePlugin` but basically, all modules do is extending nuxt and nuxt config. 

We need runtime utility from `#app`, since limitation of former is we cannot have runtime logic. Imagine for example one wants to check locale before installing a plugin.

Also agree vue 3 compostables are so powerful but yet some composables, need a companion plugin (like [vueuse/head](https://github.com/vueuse/head))

**Comment by gkweb** [comment]
Anything that makes the migration from Nuxt 2 (bridge) to Nuxt 3 easier is going to increase retention. Great proposal - I will definitely use 💯

**Comment by atinux** [comment]
After thinking of it, since we have the auto-import of plugins, this is more than enough and this might be the extra abstraction layer that we don't want.

**Reply by pi0** [reply]
Plugins are not autoimported but they are less common indeed with vue 3 and now bridge is also decoupled. Let's close this.
